urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/lp/project/LPProject.java

```diff
@@ -0,0 +1,1061 @@
+					 * However, in this case, its prevNorm is not greater than 'incorrect currNorm'.   
+					 * So we replace currNorm as nextNorm.     
+					 * 
+					 * It is also possible the 'corrct currNorm' and 'incorrect currNorm' in different quad.
+					 *       out      |/ 
+					 * ---------------|- 3
+					 *       in    1 /|  2
+					 * In this case, prevNorm=null or < 'incorrect currNorm' happens when first try to find right endpoint
+					 * And also, we can not assume nextNorm is not null      
+					 * 
+					 * If the 'incorrect currNorm' is used. The faked norm just recompute the optimal point twice. 
+					 * It does not change the reuslt because it only find optimal point. 
+					 * Just replace it with 'correct norm' or 'better incorrect norm'. (more than four line intersect on the same point. 
+					 * If nextNorm is null, jumpt to next quad.     
+					 */
+					if(prevNorm==null || prevNorm.compareTo(currNorm)<0){//fake currNorm found
+						assert nextNorm==null || nextNorm.compareTo(currNorm)>0:
+							"LPProject.findPoints: impossible. Incorrect currNorm found before, but can not find the correct currNorm";
+						//System.out.println("fake currNorm is found before, update it");
+						//currNorm is a faked norm. Ignore it and use the correct one. 
+						//And we use point, it does not change the result. Just find the optPoint twice.
+						//if nextNorm is null, jump to next quad.
+						break;
+					}
+					//prevNorm == currNorm
+					if(prevNorm.compareTo(currNorm)==0){//backward norm == currNorm ==> this basis is the right point of current edge
+//						System.out.println("The right point is found, go to next");
+						break;
+					}
+					//prevNorm > currNorm, then bisect the optimal direction and compute new basis and nextNorm
+					rOptDir = currNorm.add(prevNorm).div(2);
+				}
+
+			}
+		}
+		//project points onto plane.
+		Point[] points2D = matrix2Points(points,x_axis,y_axis,(ScaleType)x.elementType());//NOTE: convert from lpsolver type to lpproject type
+		Polygon poly = points2Polygon(points2D,ProjectFactory.errTol);		
+		Matrix vertices = poly2FullMatrix(poly,x_axis,y_axis);//x,y restore to original
+		return vertices;
+	}
+
+	/*
+	 * findPoints can find exactly result, however, it is slow. 
+	 * Some time, we can overapproximate the projected polygon by a small ammount.
+	 * Therefore, we implement findPlanes again. 
+	 * However, we don't depend on finding the correct right endpoint as before  which may introduce large error.
+	 * For each basis, we can compute the prevNorm and nextNorm, and each defines a line. 
+	 * Although some points are omitted as expected, the lines are exactly (the normal is not exactly before).
+	 * Therefore, we hope it is faster and introduce slight error. In fact, we only need to 
+	 * find all normal of edges of the reduced projected polygon.
+	 */
+	//public static double angleEps = 1e-2;
+	protected Matrix findByLine(Matrix A, Matrix b, Matrix x_axis, Matrix y_axis){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 594-624 into a separate method.